opendialogai/opendialog
# Prevent Cookies for all webchat traffic

This PR addresses the issue that users who accessed only the webchat bot of OpenDialog (ie do not log into admin pages at all) are having 2 cookies placed on their machine `XSRF_TOKEN` for CSRF checks and `{APP_NAME}_session` for the php session id. Neither of these are needed for webchat to work - the session is maintained by the user id sent with each request and CSRF checks are turned off for webchat endpoints.


The fix was to prevent the web routes in core and webchat package using the default `web` middleware group which contains the following Laravel Middlewares the we don't want:

            \App\Http\Middleware\EncryptCookies::class
            \Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse::class
            \Illuminate\Session\Middleware\StartSession::class
            // \Illuminate\Session\Middleware\AuthenticateSession::class
            \Illuminate\View\Middleware\ShareErrorsFromSession::class
            \App\Http\Middleware\VerifyCsrfToken::class

The only middleware we do want to apply (even though it's not explicitly used at the moment for any routes) is             \Illuminate\Routing\Middleware\SubstituteBindings::class - this can be referred to as `bindings` as set up [here](https://github.com/opendialogai/opendialog/blob/1.x/app/Http/Kernel.php#L56)

## How to test full page chat
- Spin up this branch and composer install
-  start the dev env
- ensure ALLOW_FULLPAGE_WEBCHAT=true is set in your `.env` file
- navigate to opendialog.test/web-chat and check for cookies.
- If there are any, clear cache and try again

## How to confirm iFrame

The steps I took to confirm this locally:

- check out this branch
- composer install
- run the dev environment
- spin up webchat-dev and server with `php artisan serve` (this could be any webpage that has a different domain)
- In webchat-dev add the following to `iframe.blade` (overwrite any exising bot scripts):

``` 
<script>
                  window.openDialogSettings = {
                    url: 'http://opendialog.test',
                    user: {
                      custom: {
                        selected_scenario: '0x77a29'
                      }
                    },
                  };
 </script>
<script src='http://opendialog.test/vendor/webchat/js/opendialog-bot.js'></script>
```

- go to http://127.0.0.1:8000/ in your browser, and go to the dev tools
- in the `Application` tab, check the cookies for http://opendialog.test. (if there are some, please clear your cache and refresh the page)

You should see cookues for 127.0.0.1:8000, but none for opendialog.test:
![image](https://user-images.githubusercontent.com/35454728/130086735-0b89bbc2-a167-4da7-a890-3dec09c8dc60.png)
